src/signature.rs

```diff
@@ -164,136 +171,295 @@ impl<const N: usize> Signature<N> {
+                                != needle.cast::<u16>().read_unaligned()
+                            {
+                                cursor +=
+                                    skip_offsets[*scan.add(N.saturating_sub(1)) as usize] as usize;
+                                continue 'outer;
+                            } else {
+                                mask = mask.add(2);
+                                needle = needle.add(2);
+                                i += 2;
+                            }
+                        }
+
+                        while i < N {
+                            if *scan.add(i) & *mask != *needle {
+                                cursor +=
+                                    skip_offsets[*scan.add(N.saturating_sub(1)) as usize] as usize;
+                                continue 'outer;
+                            } else {
+                                mask = mask.add(1);
+                                needle = needle.add(1);
+                                i += 1;
+                            }
+                        }
+
+                        let current_cursor = cursor;
+                        cursor = cursor + 1;
+                        return Some(current_cursor);
                     }
-                    let offset = skip_offsets[scan[end] as usize];
-                    current += offset as usize;
                 }
-                None
             }
-        }
+        })
+        .fuse()
     }
+}
+
+/// Trait that provides scanning methods for the `Signature` type.
+pub trait SignatureScanner {
+    /// Scans a process's memory in the given range for the first occurrence of the signature.
+    ///
+    /// # Arguments
+    ///
+    /// * `process` - A reference to the `Process` in which the scan occurs.
+    /// * `addr` - The starting address of the memory range.
+    /// * `len` - The length of the memory range to scan.
+    ///
+    /// Returns `Some(Address)` of the first match if found, otherwise `None`.
+    fn scan<'a>(&'a self, process: &Process, range: (impl Into<Address>, u64)) -> Option<Address>;
 
-    /// Scans a process for the signature. This will scan the address range of
-    /// the process given. If the signature is found, the address of the start
-    /// of the signature is returned.
-    pub fn scan_process_range(
+    /// Returns an iterator over all occurrences of the signature in the process's memory range.
+    ///
+    /// # Arguments
+    ///
+    /// * `process` - A reference to the `Process` in which the scan occurs.
+    /// * `addr` - The starting address of the memory range.
+    /// * `len` - The length of the memory range to scan.
+    ///
+    /// Returns an iterator that yields each matching address.
+    fn scan_process_range<'a>(
+        &'a self,
+        process: &'a Process,
+        range: (impl Into<Address>, u64),
+    ) -> impl Iterator<Item = Address> + 'a;
```

Comment:
I don't understand why this one returns an iterator. The name does not imply that.

Reply:
The main idea is having 2 functions, one that returns an iterator of addresses matching the signature, and another that basically stops at the first match.
I'm open to suggestions on this one, honestly.

Reply:
Sure but the name is leftover from the previous implementation which did not return an iterator. So ideally we have something like `scan_once` / `scan_iter` or so.

Reply:
I like `scan_once` / `scan_iter`. I'll refactor the code to have those functions instead.